File: src/main/java/me/aragot/hglmoderation/events/PlayerListener.java
package me.aragot.hglmoderation.events;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.connection.DisconnectEvent;
import com.velocitypowered.api.event.connection.LoginEvent;
import com.velocitypowered.api.event.player.PlayerChatEvent;
import me.aragot.hglmoderation.HGLModeration;
import me.aragot.hglmoderation.data.Notification;
import me.aragot.hglmoderation.data.PlayerData;
import me.aragot.hglmoderation.data.punishments.Punishment;
import me.aragot.hglmoderation.data.punishments.PunishmentType;
import me.aragot.hglmoderation.tools.PlayerUtils;

import java.time.Instant;
import java.util.*;

public class PlayerListener {

    public static HashMap<String, ArrayList<String>> userMessages = new HashMap<>();
    public static HashMap<String, Punishment> playerMutes = new HashMap<>();

    @Subscribe
    public void onPlayerChat(PlayerChatEvent event){

        Punishment mute = playerMutes.get(event.getPlayer().getUniqueId().toString());

        if(mute != null){
            if(mute.isActive()){
                event.setResult(PlayerChatEvent.ChatResult.denied());
                event.getPlayer().sendMessage(mute.getMuteComponent());
                return;
            }
            playerMutes.remove(event.getPlayer().getUniqueId().toString());
        }

        ArrayList<String> messages = userMessages.get(event.getPlayer().getUniqueId().toString());
        if(messages == null){
            userMessages.put(event.getPlayer().getUniqueId().toString(), new ArrayList<>(Collections.singletonList(event.getMessage())));
            return;
        }

        if(messages.size() != 15){
            messages.add(event.getMessage());
            return;
        }

        messages.remove(0);
        messages.add(event.getMessage());
    }

    @Subscribe
    public void onPlayerJoin(LoginEvent event){
        PlayerData data = PlayerData.getPlayerData(event.getPlayer());
        String hostAddress = event.getPlayer().getRemoteAddress().getAddress().getHostAddress();
        data.setLatestIp(hostAddress);
        ArrayList<Punishment> activePunishments = Punishment.getActivePunishmentsFor(data.getPlayerId(), hostAddress);

        if(!activePunishments.isEmpty()){
            for(Punishment punishment : activePunishments) {
                if(!data.getPunishments().contains(punishment.getId()))
                    data.addPunishment(punishment.getId());
                if(punishment.getTypes().contains(PunishmentType.BAN) || punishment.getTypes().contains(PunishmentType.IP_BAN)){
                    punishment.enforce(event);
                    return;
                } else if(punishment.getTypes().contains(PunishmentType.MUTE)){
                    punishment.enforce(event.getPlayer());
                }
            }
        } else if(!data.getPunishments().isEmpty() && data.getPunishmentScore() > 0){
            Punishment latest = Punishment.getPunishmentById(data.getPunishments().get(data.getPunishments().size() - 1));
            //Reset score after one year since last punishment
            if(latest.getEndsAtTimestamp() + (60 * 60 * 24 * 365) <= Instant.now().getEpochSecond()){
                data.setPunishmentScore(0);
            }
        }

        String uuid = event.getPlayer().getUniqueId().toString();
        for(Notification notif : data.getNotifications()){
            PlayerData.notificationGroups.computeIfAbsent(notif, k -> new ArrayList<>());
            PlayerData.notificationGroups.get(notif).add(uuid);
        }

    }

    @Subscribe
    public void onPlayerDisconnect(DisconnectEvent event){
        PlayerData data = PlayerData.getPlayerData(event.getPlayer());

        String uuid = event.getPlayer().getUniqueId().toString();
        for(Notification notif : data.getNotifications())
            PlayerData.notificationGroups.get(notif).remove(uuid);

        playerMutes.remove(event.getPlayer().getUniqueId().toString());
        userMessages.remove(uuid);
        HGLModeration.instance.getDatabase().updatePlayerData(data);
        PlayerData.dataList.remove(data);
        PlayerUtils.removePlayerFromCache(event.getPlayer().getUniqueId().toString());
    }

}

File: src/main/java/me/aragot/hglmoderation/events/PlayerListener.kt
package me.aragot.hglmoderation.events

import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.connection.DisconnectEvent
import com.velocitypowered.api.event.connection.LoginEvent
import com.velocitypowered.api.event.player.PlayerChatEvent
import me.aragot.hglmoderation.entity.Notification
import me.aragot.hglmoderation.entity.PlayerData
import me.aragot.hglmoderation.entity.punishments.Punishment
import me.aragot.hglmoderation.entity.punishments.PunishmentType
import me.aragot.hglmoderation.repository.PlayerDataRepository
import me.aragot.hglmoderation.repository.PunishmentRepository
import me.aragot.hglmoderation.service.player.PlayerUtils
import me.aragot.hglmoderation.service.punishment.PunishmentConverter.Companion.getMuteComponent
import me.aragot.hglmoderation.service.punishment.PunishmentManager
import java.time.Instant

class PlayerListener {

    companion object {
        var userMessages = HashMap<String, ArrayList<String>>()
        var playerMutes = HashMap<String, Punishment>()
    }

    @Subscribe
    fun onPlayerChat(event: PlayerChatEvent) {
        val mute = playerMutes[event.player.uniqueId.toString()]

        if (mute != null) {
            if (mute.isActive) {
                event.result = PlayerChatEvent.ChatResult.denied()
                event.player.sendMessage(getMuteComponent(mute))
                return
            }
            playerMutes.remove(event.player.uniqueId.toString())
        }

        val messages = userMessages[event.player.uniqueId.toString()]

        if (messages == null) {
            userMessages[event.player.uniqueId.toString()] = ArrayList(listOf(event.message))
            return
        }

        if (messages.size != 15) {
            messages.add(event.message)
            return
        }

        messages.removeAt(0)
        messages.add(event.message)
    }

    @Subscribe
    fun onPlayerJoin(event: LoginEvent) {
        val punishmentRepository = PunishmentRepository()
        val playerDataRepository = PlayerDataRepository()
        val data = playerDataRepository.getPlayerData(event.player)
        val hostAddress = event.player.remoteAddress.address.hostAddress
        data.latestIp = hostAddress

        val activePunishments = punishmentRepository.getActivePunishmentsFor(data.id, hostAddress)
        if (activePunishments.isNotEmpty()) {
            val manager = PunishmentManager()
            for (punishment in activePunishments) {
                //If Punishment is not in player data then PunishmentType == IP BAN, therefore add to player data
                if (!data.punishments.contains(punishment.id)) data.addPunishment(punishment.id)

                if (punishment.types.contains(PunishmentType.BAN) || punishment.types.contains(PunishmentType.IP_BAN)) {
                    manager.enforcePunishment(punishment, null, event)
                    return
                } else if (punishment.types.contains(PunishmentType.MUTE)) {
                    manager.enforcePunishment(punishment, event.player, null)
                }
            }
        } else if (data.punishments.isNotEmpty() && data.punishmentScore > 0) {
            val latest = punishmentRepository.getPunishmentById(data.punishments[data.punishments.size - 1])
            //Reset score after one year since last punishment
            if (latest != null && latest.endsAt + (60 * 60 * 24 * 365) <= Instant.now().epochSecond) {
                data.punishmentScore = 0
            }
        }

        val uuid = event.player.uniqueId.toString()
        for (notif in data.notifications) {
            PlayerData.notificationGroups.computeIfAbsent(
                notif
            ) { k: Notification? -> ArrayList() }
            PlayerData.notificationGroups[notif]!!.add(uuid)
        }
    }

    @Subscribe
    fun onPlayerDisconnect(event: DisconnectEvent) {
        val repository = PlayerDataRepository()
        val data = repository.getPlayerData(event.player)

        val uuid = event.player.uniqueId.toString()
        for (notif in data.notifications)
            PlayerData.notificationGroups[notif]!!.remove(uuid)

        playerMutes.remove(event.player.uniqueId.toString())
        userMessages.remove(uuid)

        repository.updateData(data)
        PlayerUtils.removePlayerFromCache(event.player.uniqueId.toString())
    }
}